### ISSUE TITLE
CallbackManager is specific to a certain callback signature and PlotObject

### ISSUE THREAD

**Issue by havocp** [bokeh/bokeh#3115]
https://github.com/bokeh/bokeh/blob/master/bokeh/util/callback_manager.py only supports callbacks that have the signature `(attr, old, new)` and are called `on_change`. 
This is only used by `PlotObject` (for example it doesn't work for `Document`, which has a different `on_change` API).

Some possibilities:
- make this a private API (underscore-prefix the module?) to be clear that it isn't general purpose
- generalize it so people can provide a callback signature and argspec, and use it in `Document` too
- move this into `PlotObject`
- make this API contain helper functions to assist in managing callbacks, but not a full mixin

In short: it feels like it should be more-fully-baked/more-general or else private for now.

Sort-of partly related: https://github.com/bokeh/bokeh/issues/3114

**Comment by bryevdv** [comment]
There are not (at this point) any callbacks on documents at all, only on models .
